modules/ServerAPI/src/server/routes/ProtocolServiceRoutes.coffee

```diff
@@ -662,153 +662,174 @@ exports.getTemplateSELFile = (req, resp) ->
-										# TODO - add try/catch 
-										if j.lsKind == "column name" and j.ignored == false
-											endpointNamesEntry = j.codeValue
-										if j.lsKind == "column units" and j.ignored == false
-											endpointUnitsEntry = j.codeValue
-										if j.lsKind == "column type" and j.ignored == false
-											endpointDataTypeEntry = j.codeValue
-										if j.lsKind == "column concentration" and j.ignored == false
-											endpointConcEntry = j.numericValue
-										if j.lsKind == "column conc units" and j.ignored == false
-											endpointConcUnitsEntry = j.codeValue
-										if j.lsKind == "column time" and j.ignored == false
-											endpointTimeEntry = j.numericValue
-										if j.lsKind == "column time units" and j.ignored == false
-											endpointTimeUnitsEntry = j.codeValue
-										if j.lsTypeAndKind == "codeValue_hide column" and j.ignored == false
-											endpointHiddenEntry = j.codeValue
-
-									# create a string of all the different sections put together to identify duplicates
-									endpointString = endpointNamesEntry + endpointUnitsEntry + endpointDataTypeEntry + String(endpointConcEntry) + endpointConcUnitsEntry + String(endpointTimeEntry) + endpointTimeUnitsEntry
-
-									# if the endpoint is not already in there, record it
-									if endpointString not in endpointStrings
-										endpointStrings.push endpointString							
-
-										# record the endpoint data
-										endpointNames.push endpointNamesEntry
-										endpointUnits.push endpointUnitsEntry
-										endpointDataTypes.push endpointDataTypeEntry
-										endpointConcs.push endpointConcEntry
-										endpointConcUnits.push endpointConcUnitsEntry
-										endpointTimes.push endpointTimeEntry
-										endpointTimeUnits.push endpointTimeUnitsEntry
-										endpointHiddens.push endpointHiddenEntry
+						@endpointNames = []
+						@endpointUnits = []
+						@endpointDataTypes = []
+						@endpointConcs = []
+						@endpointConcUnits = []
+						@endpointTimes = []
+						@endpointTimeUnits = []
+						@endpointHiddens = []
+
+						@endpointStrings = []
+
+						# set the default value to true unless otherwise specified
+						protocolHasEndpoints = true
+
+						# if the protocol has no experiments, use the defined endpoint information in the protocol 
+						if experiments.length == 0
+							protocolEndpointLsStates = _.where(json.lsStates, {lsKind: "data column order", ignored: false})
+							# if the protocol has no endpoints saved in it, we record it has no endpoint data and won't write that section of the .csv...
+							if protocolEndpointLsStates.length == 0
+								protocolHasEndpoints = false 
+
+							# ...but if it does, we want to extract them and format them for the template file
+							else
+								for lsState in protocolEndpointLsStates
+									extractEndpointData(lsState)
+
+						# if the protocol has experiments, collect all the endpoint information available
```

Comment:
@hhan-schrodinger @brianbolt Do we expect the experiment endpoints to differ from the protocol endpoints? If so, wouldn't we want the template to reflect the curated endpoints defined on the protocol?

I don't see a reason to check the experiments at all - rather I think we shuld just be reading the `protocolEndPointLsStates`. Can either of you think of a reason why we should be looking at the experiments? It's possible I'm missing something here.

Reply:
If you create a protocol when uploading an experiment via Experiment Loader (SEL) does it create the protocolEndPointLsStates?  If not, then we should add it and then I don't think there is every a reason to look at the experiment level when using the protocol editor.

Reply:
I think that is much simpler to have the template just reflect endpoints defined on the protocol too. I vaguely remember looking across all the endpoints in all associated experiments being the original intention maybe to find all possible endpoints for a protocol (maybe to look for erroneous endpoints that wouldn't show up in the endpoint manager)?

Reply:
@hhan-schrodinger Let's simplify it just to look at the protocol endpoints. Less code means less places for bugs to hide.

Reply:
@bffrost Sounds good, made the changes!